catkit2/services/accufiz_interferometer/accufiz_interferometer.py

```diff
@@ -0,0 +1,196 @@
+
+        self.log.info(f"{self.config_id}: Succeeded to save measurement data to '{local_file_path}'")
+
+        mask = np.array(h5py.File(local_file_path, 'r').get('measurement0').get('Detectormask', 1))
+        img = np.array(h5py.File(local_file_path, 'r').get('measurement0').get('genraw').get('data')) * mask
+
+        self.detector_masks.submit_data(mask.astype(np.uint8))
+
+        image = self.convert_h5_to_fits(local_file_path, rotate=0, fliplr=True, mask=mask, img=img, create_fits=self.save_fits)
+        # remove h5 file if configuration was not set
+        if (not self.save_h5) and os.path.exists(local_file_path):
+            os.remove(local_file_path)
+
+        return image
+
+
+    @staticmethod
+    def convert_h5_to_fits(filepath, rotate, fliplr, img, mask, wavelength=632.8, create_fits=False):
+
+        filepath = filepath if filepath.endswith(".h5") else f"{filepath}.h5"
+
+        fits_filepath = f"{os.path.splitext(filepath)[0]}.fits"
+
+        mask = np.array(h5py.File(filepath, 'r').get('measurement0').get('Detectormask', 1))
+        img = np.array(h5py.File(filepath, 'r').get('measurement0').get('genraw').get('data')) * mask
+        if create_fits:
+            fits.PrimaryHDU(mask).writeto(fits_filepath, overwrite=True)
+
+        radiusmask = np.int64(np.sqrt(np.sum(mask) / math.pi))
+        center = ndimage.measurements.center_of_mass(mask)
+
+        image = np.clip(img, -10, +10)[np.int64(center[0]) - radiusmask:np.int64(center[0]) + radiusmask - 1,
+                                       np.int64(center[1]) - radiusmask: np.int64(center[1]) + radiusmask - 1]
+
+        # Apply the rotation and flips.
+        image = rotate_and_flip_image(image, rotate, fliplr)
+
+        # Convert waves to nanometers.
+        image = image * wavelength
+        if create_fits:
+            fits_hdu = fits.PrimaryHDU(image)
+            fits_hdu.writeto(fits_filepath, overwrite=True)
+        return image
+
+    def main(self):
+        while not self.should_shut_down:
+            if self.should_be_acquiring.wait(0.05):
+                self.acquisition_loop()
+
+    def acquisition_loop(self):
+        try:
+            self.is_acquiring.submit_data(np.array([1], dtype='int8'))
+
+            while self.should_be_acquiring.is_set() and not self.should_shut_down:
+                img = self.take_measurement()
+
+                # Make sure the data stream has the right size and datatype.
+                has_correct_parameters = np.allclose(self.images.shape, img.shape)
+
+                if not has_correct_parameters:
+                    self.images.update_parameters('uint16', img.shape, 20)
+                self.log.info('requesting new frame')
+                frame = self.images.request_new_frame()
+
+                frame.data[:] = img
+                self.log.info(f'frame {frame.id} acquired')
+                self.images.submit_frame(frame.id)
+                time.sleep(0.05)
```

Comment:
That procedure to submit into a datastream is correct. But that's very low level. Unless there's a specific reason to use that, you can use the more convenient `self.images.submit_data()` function. That will create a new frame and submit it automatically.

Reply:
Oh yes, the dummy camera is probably one of the early piece of code written by Emiel. New things have been added since then 😅

Reply:
The `zwo_camera.py` service is probably more up to date.